src/findex.rs

```diff
@@ -0,0 +1,262 @@
+use std::{
+    collections::{HashMap, HashSet},
+    hash::Hash,
+    sync::{Arc, Mutex},
+};
+
+use tiny_keccak::{Hasher, Sha3};
+
+use crate::{
+    ADDRESS_LENGTH, Address, IndexADT, KEY_LENGTH, MemoryADT, adt::VectorADT, encoding::Op,
+    encryption_layer::MemoryEncryptionLayer, error::Error, ovec::IVec, secret::Secret,
+};
+
+#[derive(Clone, Debug)]
+pub struct Findex<
+    const WORD_LENGTH: usize,
+    Value,
+    TryFromError: std::error::Error,
+    Memory: Send + Sync + Clone + MemoryADT<Address = Address<ADDRESS_LENGTH>, Word = [u8; WORD_LENGTH]>,
+> where
+    // values are serializable (but do not depend on `serde`)
+    for<'z> Value: TryFrom<&'z [u8], Error = TryFromError> + AsRef<[u8]>,
+    Memory::Error: Send + Sync,
+{
+    el: MemoryEncryptionLayer<WORD_LENGTH, Memory>,
+    cache: Arc<
+        Mutex<
+            HashMap<
+                Address<ADDRESS_LENGTH>,
+                IVec<WORD_LENGTH, MemoryEncryptionLayer<WORD_LENGTH, Memory>>,
+            >,
+        >,
+    >,
+    encode: Arc<
+        fn(
+            Op,
+            HashSet<Value>,
+        )
+            -> Result<Vec<<MemoryEncryptionLayer<WORD_LENGTH, Memory> as MemoryADT>::Word>, String>,
+    >,
+    decode: Arc<
+        fn(
+            Vec<<MemoryEncryptionLayer<WORD_LENGTH, Memory> as MemoryADT>::Word>,
+        ) -> Result<HashSet<Value>, TryFromError>,
+    >,
+}
+
+impl<
+    const WORD_LENGTH: usize,
+    Value: Send + Sync + Hash + Eq,
+    TryFromError: std::error::Error,
+    Memory: Send + Sync + Clone + MemoryADT<Address = Address<ADDRESS_LENGTH>, Word = [u8; WORD_LENGTH]>,
+> Findex<WORD_LENGTH, Value, TryFromError, Memory>
+where
+    for<'z> Value: TryFrom<&'z [u8], Error = TryFromError> + AsRef<[u8]>,
+    Vec<u8>: From<Value>,
+    Memory::Error: Send + Sync,
+{
+    /// Instantiates Findex with the given seed, and memory.
+    pub fn new(
+        seed: Secret<KEY_LENGTH>,
+        mem: Memory,
+        encode: fn(Op, HashSet<Value>) -> Result<Vec<[u8; WORD_LENGTH]>, String>,
+        decode: fn(Vec<[u8; WORD_LENGTH]>) -> Result<HashSet<Value>, TryFromError>,
```

Comment:
I feel these 2 callbacks make Findex instantiation hard (and the temptations to reuse the the dummy_ functions high)

Reply:
Maybe, but they allow fitting the encoding to the specific data used. For example, if only storing 16B values, the overhead is virtually null (should be only 1bit per value). If storing entire files it can be adjusted to perform compression plus storing big chunks of data at once (still some overhead but it can be reduced).

This differs from Findex v6 in which a specific encoding strategy was hard-coded. It was optimized to store 16B data so storing entire files incurred a non-negligible overhead.